File: bin/phpesme_service_example.rb
#!/usr/bin/ruby
# coding: utf-8

require 'mysql'

DB_HOST = 'localhost'
DB_USER = 'user'
DB_PASS = 'password'
DB_SCHEME = 'phpesme'

$dbh = nil

def mysql_is_down?
  begin
    $dbh.query('SELECT 1')
    return false
  rescue
    $dbh = Mysql.real_connect(DB_HOST, DB_USER, DB_PASS, DB_SCHEME)
    if $dbh
      $dbh.charset = 'utf8'
      $dbh.autocommit(false)
      return false
    else
      return true
    end
  end
end

Signal.trap(3) { $dbh.close if $dbh }
Signal.trap(2) { $dbh.close if $dbh }

while true do # service always running
  if mysql_is_down?
    sleep 5
    next
  end
  
  $dbh.commit # this is required to get non-cached result
  $dbh.query('SELECT id,src,dst,msg,ts FROM inbox WHERE mt=0 AND dst LIKE "3333.%"').each_hash do |inbox_msg| # processing incoming messages one by one
    
    reply_text = inbox_msg['msg'].strip.reverse # will reply with original text reversed
    
    begin
      # creating message in outbox
      $dbh.query("INSERT INTO outbox (rel_id, src, dst, msg, ts, rd)
                 VALUES (#{inbox_msg['id']}, '3333.0.1', '#{inbox_msg['src']}','#{$dbh.escape_string(reply_text)}',#{Time.now.to_i},0)")
      # putting original message in archive, providing sent message id as resp_id
      $dbh.query("INSERT INTO archive(id,src,dst,msg,ts,proc_ts,resp_id) VALUES(#{inbox_msg['id']},\"#{inbox_msg['src']}\",\"#{inbox_msg['dst']}\",\"#{$dbh.escape_string(inbox_msg['msg'])}\",#{inbox_msg['ts']},#{Time.now.to_i},#{$dbh.insert_id})")
      # removing original message from inbox
      $dbh.query("DELETE FROM inbox WHERE id=#{inbox_msg['id']}")
      $dbh.commit
    rescue 
      $dbh.rollback
      break
    end
    
  end
  sleep 5
end

$dbh.close if $dbh

File: bin/phpesme_service_example.php
#!/usr/bin/php -q
<?php

$conn = null;

define('MYDIR',substr(dirname(__FILE__),0,-4));

$conf = parse_ini_file(MYDIR.'/conf/phpesme.conf', true);

require(MYDIR.'/lib/nimf_mysqli.php');

mb_internal_encoding("UTF-8");

declare(ticks = 1);
pcntl_signal(SIGTERM, "sig_handler");
pcntl_signal(SIGHUP, "sig_handler");
pcntl_signal(SIGINT, "sig_handler");
pcntl_signal(SIGQUIT, "sig_handler");
pcntl_signal(SIGABRT, "sig_handler");

while (true) {
  
  if (!isset($db)) {
    $db = new NIMF_mysqli($conf['DB']['host'], $conf['DB']['login'], $conf['DB']['pass'], $conf['DB']['db']);
    if (mysqli_connect_errno()) {
      unset($db);
      sleep(5);
      continue;
    } else {
      $db->put('SET NAMES "'.$conf['DB']['charset'].'"');
      $db->put('SET SESSION query_cache_type = OFF');
      $db->autocommit(false);
    }
  }
  
  // Get messages
  $db->commit(); // <-- this is required for getting non-cached result
  $in = $db->get_array('SELECT id,src,dst,msg,ts FROM inbox WHERE mt=0 AND dst LIKE "3333.%"');
  
  if (isset($in['id'])) {
    
    // Process messages
    foreach ($in['id'] as $k=>$inbox_msg_id) {
      
      $reply = strrev(trim($in['msg'][$k])); // will reply with original text reversed
      
      $time = time();
      // creating message in outbox
      $db->put('INSERT INTO outbox(rel_id,src,dst,msg,ts,rd) VALUES('.$inbox_msg_id.',"'.$in['dst'][$k].'","'.$in['src'][$k].'","'.$db->real_escape_string($reply).'",'.$time.',0)');
      // putting original message in archive, providing sent message id as resp_id
      $db->put('INSERT INTO archive(id,src,dst,msg,ts,proc_ts,resp_id) VALUES('.$inbox_msg_id.',"'.$in['src'][$k].'","'.$in['dst'][$k].'","'.$db->real_escape_string($in['msg'][$k]).'",'.$in['ts'][$k].','.$time.','.$db->get_id().')');
      // removing original message from inbox
      $db->put('DELETE FROM inbox WHERE id='.$inbox_msg_id);
      if (!$db->commit()) {
        unset($db); break;
      }
      
    }
    
  }
  
  sleep(5);
} // infinite loop

function sig_handler($signo) {
  global $db;
  switch ($signo) {
    case SIGTERM:
    case SIGHUP:
    case SIGINT:
    case SIGQUIT:
    case SIGABRT:
      if (isset($db)) $db->close();
      die();
    break;
    default:
     return true;
  }
}
